### ISSUE TITLE
DNS resolution of other containers on network broken after 24.0.3

### ISSUE THREAD

**Issue by MattJud** [moby/moby#45928]
### Description

I've recently upgraded my docker engine from 24.0.2 -> 24.0.4. However, I'm encountering an error with the docker namespace resolver. The container appears unable to find the namespace of the other container, but I'm not sure yet why, as it was working fine before the upgrade. 

Furthermore, this error looks pretty random since if I restart my stack, the problem will occur on a different container, or if I'm lucky, everything works fine.

I currently found a way to make my stack working again with docker engine 24.0.4 by running `sudo systemctl restart docker`. After doing this, my stack can run normally again. But if I rebuild my containers again, this problem will occur again and I need to restart the docker instance again. I've also tried downgrading to 24.0.2, and everything seems normal. All containers can immediately resolve the namespace even after rebuilding.

I've tested this behavior in 2 different local test deployment systems, and both show the same symptom. I don't know precisely what is causing this. If anyone can help me with debugging this, you're more than welcome to give me suggestions if I'm doing something wrong.

### Reproduce

1. Use docker engine 24.0.4
2. Rebuild containers
3. See the status of containers

### Expected behavior

All containers can find the namespace of the other container if they're in the same network

### docker version

```bash
Client: Docker Engine - Community
 Version:           24.0.4
 API version:       1.43
 Go version:        go1.20.5
 Git commit:        3713ee1
 Built:             Fri Jul  7 14:50:57 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.4
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.5
  Git commit:       4ffc614
  Built:            Fri Jul  7 14:50:57 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.21
  GitCommit:        3dce8eb055cbb6872793272b4f20ed16117344f8
 runc:
  Version:          1.1.7
  GitCommit:        v1.1.7-0-g860f061
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    24.0.4
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.19.1
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 16
  Running: 15
  Paused: 0
  Stopped: 1
 Images: 83
 Server Version: 24.0.4
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
 Kernel Version: 5.15.0-76-generic
 Operating System: Ubuntu 20.04.6 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.34GiB
 Name: matthew-X1-Carbon
 ID: IAVW:O365:N5QE:4EHC:IRXZ:UOFS:DRBU:RR7K:NSCD:GYUJ:R4Z6:HUXG
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by DaMandal0rian** [comment]
Also confirm this is happening for `version 24.0.4`

**Comment by neersighted** [comment]
What is a 'namespace' in this context? Do you mean DNS name? Likewise, you referred to a stack, but you are not using Swarm -- do you mean a Compose file?

**Comment by neersighted** [comment]
Thank you for the clarification! Skimming the delta between these tags (https://github.com/moby/moby/compare/v24.0.2...v24.0.4) there isn't anything that jumps out as a possible cause of regressions, but I'll ask @corhere and @akerouanton who have been working in this area recently to give it a more expert look.

**Comment by MattJud** [comment]
The interesting aspect of the issue is that I couldn't consistently reproduce the same error on every restart. The container that has the problem just randomly changes at every restart. And if I'm lucky, all containers will just behave normally. Also, quite an interesting observation that I see sometimes is that I have two containers running the same image with slightly different parameters. One of them can resolve the namespace to the webserver and the other couldn't.

**Comment by DaMandal0rian** [comment]
> 

Yes for me the API container can't connect to the db container when using the db service name, and all containers are running in same bridge network.

**Comment by thaJeztah** [comment]
Thanks for the reproduction repo, that's really useful!

@corhere mentioned on Slack that he managed to reproduce, but we haven't found the root cause yet

**Comment by DaMandal0rian** [comment]
I confirm a rollback of docker compose to version `2.18.1` resolved the issue.

**Comment by MattJud** [comment]
Thanks for the help! For now, I will downgrade my compose to version `2.18.1` too. Any idea when we will have another release? Would be nice to have the new Compose 2.20.0

**Comment by thaJeztah** [comment]
We haven't planned a release yet, but may be doing one once we have https://github.com/moby/moby/issues/45935 settled

**Comment by thaJeztah** [comment]
That one may need a new release of compose though (with the fix for the above issue included, so an update version of the docker engine API client)

/cc @milas

**Comment by neersighted** [comment]
The version of Compose with the fix merged is release (2.20.0).
